File: Amazon/152.maximum-product-subarray.java
/**
 * Algorithm:
 * Track both max and min product ending at the current position because a
 * negative number swaps their roles. The answer is the best max seen.
 *
 * Complexity:
 * Time O(n), space O(1).
 */
class Solution {
    public int maxProduct(int[] nums) {
        int curMax = nums[0];
        int curMin = nums[0];
        int ans = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            if (num < 0) {
                int tmp = curMax;
                curMax = curMin;
                curMin = tmp;
            }
            curMax = Math.max(num, curMax * num);
            curMin = Math.min(num, curMin * num);
            ans = Math.max(ans, curMax);
        }
        return ans;
    }
}

File: Infosys/152.maximum-product-subarray.py
"""
Approach: Kadane-style DP tracking both max and min product ending at the current index.
Data structure: two scalar states are enough because only subarrays ending at the previous element can extend to the current element.
Interview logic: a negative number swaps the role of previous maximum and minimum. Zero naturally resets because starting fresh at num is always considered.
Complexity: O(n) time, O(1) space.
Tests and edge cases: one element; zeros between segments; even vs odd count of negative values.
"""
from __future__ import annotations
from typing import List

# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        cur_max = cur_min = ans = nums[0]
        for num in nums[1:]:
            if num < 0:
                cur_max, cur_min = cur_min, cur_max
            cur_max = max(num, cur_max * num)
            cur_min = min(num, cur_min * num)
            ans = max(ans, cur_max)
        return ans
# @lc code=end
